src/internal/request-batcher.ts

```diff
@@ -0,0 +1,126 @@
+import { ConnectionInfo, FetchJsonResponse } from '@ethersproject/web';
+
+import { DEFAULT_MAX_REQUEST_BATCH_SIZE } from '../util/const';
+import { JsonRpcRequest } from './internal-types';
+
+/**
+ * Internal class to enqueue requests and automatically send/process batches.
+ *
+ * The underlying batching mechanism is loosely based on ethers.js's
+ * `JsonRpcBatchProvider`.
+ *
+ * @internal
+ */
+export class RequestBatcher {
+  /** Timeout timer that periodically sends the pending batch. */
+  private pendingBatchTimer: NodeJS.Timer | undefined;
+
+  /**
+   * Array of enqueued requests along with the constructed promise handlers for
+   * each request.
+   */
+  private pendingBatch: Array<BatchRequest> | undefined;
+
+  constructor(
+    private readonly sendBatchFn: SendBatchFn,
+    private readonly connection: ConnectionInfo,
```

Comment:
Instead of having this class take a complicated `SendBatchFn` as well as `ConnectionInfo`, I'd consider just having one param which is `(req: JsonRpcRequest) => Promise<JsonRpcResponse>`, and then the caller can be responsible for crafting that function to pass in. My reasoning is that passing the connection info around, serializing JSON, and converting the response are not related to batching at all. Further, it's likely that the class calling this (the provider) also needs this logic for the nonbatched case, and so as written you get some code duplication between this class and the caller.

Reply:
Oh whoops, I just realized I forgot to make the request and response arrays in the type I put in above. Thanks for figuring out what I meant to say!